### PR TITLE
Provide editing capabilities using Mono.Terminal.GetLine.

### PR DESCRIPTION
**Pull request by migueldeicaza**
This provides command line editing, history, and reverse history
search using the Emacs keybindings, similar to bash.

### CODE REVIEW

**Comment by migueldeicaza** [comment]
The line editor also contains support for code completion, to hook it up, you should follow the steps that are done in this file:

https://github.com/mono/mono/tree/master/mcs/tools/csharp

**Comment by glennblock** [comment]
Awesome!

File: src/ScriptCs.Contracts/IConsole.cs [hunk 1]
```diff
@@ -10,7 +10,7 @@ public interface IConsole
 
         void WriteLine(string value);
 
-        string ReadLine();
+        string ReadLine(string prompt);
```
Comment (adamralph) [inline src/ScriptCs.Contracts/IConsole.cs]: I wonder if we can avoid this breaking change. I'll have a play...

Comment (glennblock) [inline src/ScriptCs.Contracts/IConsole.cs]: Good point. The code we had before should work / we don't need prompt to be a param.

File: src/ScriptCs.Hosting/ScriptConsole.cs [hunk 2]
```diff
@@ -1,13 +1,15 @@
...
 
 namespace ScriptCs.Hosting
 {
     public class ScriptConsole : IConsole
     {
+		LineEditor le;
         public ScriptConsole()
         {
-            Console.CancelKeyPress += HandleCancelKeyPress;
+			le = new LineEditor ("scriptcs");
```
Comment (glennblock) [inline src/ScriptCs.Hosting/ScriptConsole.cs]: Can we reformat the code? :-)

File: src/ScriptCs.Hosting/ScriptCs.Hosting.csproj [hunk 3]
```diff
@@ -63,6 +51,15 @@
     <Reference Include="Microsoft.CSharp" />
     <Reference Include="System.Data" />
     <Reference Include="System.Xml" />
+    <Reference Include="Autofac">
```
Comment (glennblock) [inline src/ScriptCs.Hosting/ScriptCs.Hosting.csproj]: We should revert dependency changes / keep the old packages.

File: src/ScriptCs.Hosting/getline.cs [hunk 4]
```diff
@@ -0,0 +1,1086 @@
+//
+// getline.cs: A command line editor
```
Comment (glennblock) [inline src/ScriptCs.Hosting/getline.cs]: Can we make this GetLine.cs?

File: src/ScriptCs.Hosting/getline.cs [hunk 5]
```diff
@@ -0,0 +1,1086 @@
...
+//    System.Console needs to get the DELETE character, and report accordingly.
+//
+
+using System;
+using System.Text;
+using System.IO;
+using System.Threading;
+using System.Reflection;
+
+namespace Mono.Terminal {
```
Comment (glennblock) [inline src/ScriptCs.Hosting/getline.cs]: formatting?

**Comment by glennblock** [comment]
@migueldeicaza, thank you for this! We're going to close and pull your branch and do a little cleanup and then we'll push your commits as well as some tweaks.

**Comment by adamralph** [comment]
We don't have to close. We can add clean up commit(s), send a new PR, merge it and then this one will show as merged too.

**Comment by glennblock** [comment]
I got this folks, I'll clean it up and add unit tests.

**Comment by glennblock** [comment]
Let's add Miguel to the contributors list! Thank you @migueldeicaza!

**Comment by adamralph** [comment]
Subset of #1119

**Pull request merged**